Add a vmware-dev variant

zmrow opened the pull request:
**Issue number:**
N/A


**Description of changes:**
This change adds the necessary files for a VMWare development variant.  The files are currently a copy of the `aws-dev` variant but will diverge over time.

This variant currently includes some AWS settings that are required to fully boot, but those settings will slowly disappear as we eliminate any AWS dependencies in the software.


**Testing done:**
* Build and run the variant locally (including the changes from #1288 and #1290 ) to ensure that the variant boots and correctly looks at `/media/cdrom` for userdata


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

zmrow committed:
Add a vmware-dev variant

This change adds the necessary files for a VMWare development variant.
The files are currently a copy of the aws-dev variant but will diverge
over time.

samuelkarp commented on sources/models/src/vmware-dev/mod.rs:
```diff
@@ -0,0 +1,21 @@
+    ntp: NtpSettings,
+    network: NetworkSettings,
+    kernel: KernelSettings,
+    aws: AwsSettings,
```
might be worth a comment that we intend to remove this setting

zmrow replied:
Per some offline conversation - we removed this completely.  :) 

Even though some system services end up failing because it's missing, it's probably important to remove the crutch early and work towards fixing whatever is still broken rather than leaving it in and having it be a pain to dig out later.

tjkirch replied:
I'm hesitant to support merging a new variant that's known to be broken.  We're definitely working toward removing this crutch, and it's easy to test what happens if we remove it...

zmrow replied:
I can see where you're coming from, but this variant is to facilitate development.  It's hard to support fixing the variant if there's not a variant to build. :)

tjkirch replied:
> It's hard to support fixing the variant if there's not a variant to build.

I don't understand.  The new variant would be available, would have `settings.aws` so it's not broken, and development could continue toward removing `settings.aws`.

tjkirch commented on sources/models/src/vmware-dev/override-defaults.toml:
```diff
@@ -0,0 +1,12 @@
+[configuration-files.containerd-config-toml]
+# No override to path
+template-path = "/usr/share/templates/containerd-config-toml_aws-dev"
```
Can we make a vmware-dev file instead?  If it's the same, it could be a symlink in the packages directory, but having the wrong variant name on the system at runtime feels wrong...

zmrow replied:
I agree.  A symlink seems fine.

zmrow replied:
I couldn't use a symlink for this file and still reference it in the `containerd.spec` so I opted to just create an actual file.

tjkirch replied:
We should be able to use a symlink; what happened when you tried?  Can we dig in a bit?  If you post or share the changes you tried, I bet we can figure it out.

bcressey replied:
Either way we should update the path here to point to `containerd-config-toml_vmware-dev`.

tjkirch commented on sources/logdog/conf/logdog.vmware-dev.conf:
```diff
@@ -0,0 +1,2 @@
+exec docker-info docker info
```
I think we could have this be a link to the aws-dev file to reflect that we expect no differences.  (We don't expect any, right?)

zmrow replied:
I'll symlink it for now.  If we find differences later, we'll rectify it then.

tjkirch reviewed (commented):
I think we should add this to `.github/workflows/build.yml`, as an optional build like aws-dev, to make sure we don't break it as we develop.

I guess we can wait on `sources/models/src/lib.rs` (for README generation) and `variants/README.md` until there's some useful difference, but let's not forget...

[edit] Actually, the cdrom difference is there, so it might be worth doing this now?

zmrow commented:
> I guess we can wait on sources/models/src/lib.rs (for README generation) and variants/README.md until there's some useful difference, but let's not forget...
[edit] Actually, the cdrom difference is there, so it might be worth doing this now?

Good catch - I completely missed this.

zmrow commented:
^ Addressed all the comments!

tjkirch commented on sources/models/src/lib.rs:
```diff
@@ -52,6 +52,11 @@ The `#[model]` attribute on Settings and its sub-structs reduces duplication and
 * [Model](src/aws-dev/mod.rs)
```
Can we also update the aws-dev header to say "AWS development build" to clarify the distinction?  It won't let me make a suggestion.

Same in `variants/README.md` - the headers are currently the same there, should probably make it match this wording.  And in that file, there are clarifications for VMWare that didn't make it back into the aws-dev paragraph.  I think it'd be cleaner to have each describe its setup, instead of only saying the differences in the VMWare entry.

tjkirch commented on sources/models/src/lib.rs:
```diff
@@ -52,6 +52,11 @@ The `#[model]` attribute on Settings and its sub-structs reduces duplication and
 * [Model](src/aws-dev/mod.rs)
 * [Overridden defaults](src/aws-dev/override-defaults.toml)
 
+## vmware-dev: VMWare Development build
```
```suggestion
## vmware-dev: VMWare development build
```

tjkirch commented on variants/README.md:
```diff
@@ -69,6 +69,12 @@ container instance in AWS.
 
+### vmware-dev: Development build
+
+The [vmware-dev](vmware-dev/Cargo.toml) variant has useful packages for local development of the OS for running as a VMWare guest.
```
```suggestion
The [vmware-dev](vmware-dev/Cargo.toml) variant has useful packages for local development of the OS, and is intended to run as a VMWare guest.
```

tjkirch commented on variants/README.md:
```diff
@@ -69,6 +69,12 @@ container instance in AWS.
+### vmware-dev: Development build
+
+The [vmware-dev](vmware-dev/Cargo.toml) variant has useful packages for local development of the OS for running as a VMWare guest.
+It includes tools for troubleshooting as well as Docker for running containers, and doesn't depend on AWS or IMDS for things like user data.
+User data will be read from a mounted CD-ROM, either from a file named "user-data" or from an OVF file.
```
```suggestion
It includes tools for troubleshooting as well as Docker for running containers.
User data will be read from a mounted CD-ROM, either from a file named "user-data" or from an OVF file.
```

zmrow commented:
^ Addressed @tjkirch 's comments

bcressey commented on sources/models/src/vmware-dev/override-defaults.toml:
```diff
@@ -0,0 +1,12 @@
+restart-commands = ["/bin/systemctl try-restart docker.service"]
+configuration-files = ["proxy-env"]
+
+# Network
+[metadata.settings.network]
+affected-services = ["containerd", "docker", "host-containerd"]
```
Should we also override the default host containers, since they aren't expected to work?

zmrow committed:
host-containers: Don't fail if host containers aren't defined

If a user doesn't define any host containers, it only means that this
program has nothing to do.  This change prevents the program from
failing, which causes other boot services to fail.

zmrow commented:
^ The above push:
* Addresses the remaining comments
* Takes advantage of the default settings refactor from #1303 
* Add a commit that fixes a small bug in `host-containers`.  Previously if a user didn't define any host containers, the program would fail.  The change ensures that the program continues on gracefully since it has nothing to do.

tjkirch commented on sources/models/src/vmware-dev/defaults.d/50-vmware-dev.toml:
```diff
@@ -0,0 +1,18 @@
+affected-services = ["containerd", "docker", "host-containerd"]
+
+# NTP
+# Override the 169. endpoint as we can't trust it off of AWS
```
```suggestion
# Use a public endpoint, don't assume any local ones.
```

zmrow commented:
^ Integrate @tjkirch 's wording suggestion
